src/dotnet/Chat.Service/Places.cs

```diff
@@ -95,19 +107,50 @@ public virtual async Task<ApiArray<AuthorId>> ListOwnerIds(Session session, Plac
     public virtual async Task<Place> OnChange(Places_Change command, CancellationToken cancellationToken)
     {
         var (session, placeId, expectedVersion, placeChange) = command;
-        if (placeChange.Remove)
-            throw StandardError.Constraint("Use Places_Delete command instead to delete the place.");
 
-        var chatChange = new Change<ChatDiff> {
-            Create = placeChange.Create.HasValue ? Option<ChatDiff>.Some(ToChatDiff(placeChange.Create.Value)) : default,
-            Update = placeChange.Update.HasValue ? Option<ChatDiff>.Some(ToChatDiff(placeChange.Update.Value)) : default,
-            Remove = placeChange.Remove
-        };
-        var chatId = placeId.ToRootChatId();
-        var chatChangeCommand = new Chats_Change(session, chatId, expectedVersion, chatChange);
+        var place = placeId.IsNone ? null
+            : await Get(session, placeId, cancellationToken).ConfigureAwait(false);
+
+        var changePlaceCommand = new PlacesBackend_Change(placeId, expectedVersion, placeChange.RequireValid());
+        ChatId rootChatId;
+        long? chatExpectedVersion;
+        Change<ChatDiff> chatDiff;
+        Chat? placeRootChat;
+        if (placeChange.Kind == ChangeKind.Create) {
+            var account = await Accounts.GetOwn(session, cancellationToken).ConfigureAwait(false);
+            account.Require(AccountFull.MustBeActive);
+            rootChatId = ChatId.None;
+            chatExpectedVersion = null;
+            chatDiff = Change.Create(ToChatDiff(placeChange.Create.Value));
+        }
+        else {
+            var requiredPermissions = placeChange.Remove
+                ? PlacePermissions.Owner
+                : PlacePermissions.EditProperties;
+            place.Require().Rules.Permissions.Require(requiredPermissions);
+
+            if (placeChange.Remove)
+                await RemovePlaceChats(session, placeId, cancellationToken).ConfigureAwait(false);
+
+            rootChatId = placeId.ToRootChatId();
+            placeRootChat = await Chats.Get(session, rootChatId, cancellationToken).ConfigureAwait(false);
+            placeRootChat.Require();
+            chatExpectedVersion = placeRootChat.Version;
+            chatDiff = !placeChange.Remove ? Change.Update(ToChatDiff(placeChange.Update.Value)) : Change.Remove<ChatDiff>();
+        }
+
+        place = await Commander.Call(changePlaceCommand, cancellationToken).ConfigureAwait(false);
+        rootChatId = rootChatId.Or(place.Id.ToRootChatId());
+
+        // Since place root chat is still used for getting place rules and place members we should synchronously update it.
+        // Should we use backend command here?
+        var chatChangeCommand = new Chats_Change(session, rootChatId, chatExpectedVersion, chatDiff);
+        placeRootChat = await Commander.Call(chatChangeCommand, true, cancellationToken).ConfigureAwait(false);
```

Comment:
Why do you use outermost command there?

Reply:
Usually we prefer backend command calls

Reply:
Probably we should move this call to the Places backend so it can incapsulate all place-chat sync logic

Reply:
Reworked